### ISSUE TITLE
ignore_index (and similar) aren't rewriting the query

### ISSUE THREAD

**Issue by kinghelix** [adamchainz/django-mysql#677]
Hi, I'm trying to use ignore_index() from django-mysql to stop mysql from incorrectly using an inefficient index (and use a composite index instead etc...)

I've got:

```
pip:
Django==2.2.6
django-mysql==3.7.0
mysqlclient==1.4.4

MySQL 5.6.47 via Brew
macOS High Sierra
```
```
# in settings:

INSTALLED_APPS.append("django_mysql")
DJANGO_MYSQL_REWRITE_QUERIES = True
```

```
# in the view:
from django_mysql.models import add_QuerySetMixin
results = FooBar.objects.all()
results = results.filter(foobar=somevariable)
results = add_QuerySetMixin(results)
results = results.ignore_index("main_foobar_created_at_abcd12345_uniq")
print(results.query)
```

then when I iterate over `results` or print the query, it doesn't work, instead I can see the commented hint which apparently never gets handled and the sql never gets the IGNORE INDEX stuff added:

```
print(results.query)
[...] AND (/*QueryRewrite':index=`main_foobar` IGNORE `main_foobar_created_at_abcd12345_uniq`*/1))
```

Also, when i do `results.explain()` it confirms that the index I want to ignore isn't being ignored.

I'm using approach number 4 in the installation docs (use `add_QuerySetMixin`) but I've tried the other approaches as well and I still have the same results. I've also tried `force_index` and `use_index` and they have the same results.

Am I missing something? Is it broken? Thanks very much! :)

**Comment by adamchainz** [comment]
How are your settings laid out? If you're running the `append()` at some point after django has first read the `INSTALLED_APPS` setting, it might not actually be installed, which stops the query rewriting hook being installed.

**Comment by kinghelix** [comment]
Ah sorry, normally I've added it as a part of the initial `INSTALLED_APPS` list that's at the top of `settings.py` (the normal way you're supposed to do it). I also experimented with doing it as an `append()` which is what I pasted into this ticket.

Either way that I did it, I still got the same result. I've just added it as a part of `INSTALLED_APPS` and it's still not rewriting the query.

**Comment by adamchainz** [comment]
Can you check with version 3.5.0 as 3.6.0 chagned the way query rewriting worked? https://github.com/adamchainz/django-mysql/blob/master/HISTORY.rst#360-2020-06-09

**Comment by kinghelix** [comment]
Hmm ok this is interesting, so as you suggested, I downgraded from 3.7.0 to 3.5.0 and now it's working.

However, when I print result.query, I still don't see the rewritten query - it still has the commented hint in it.

Also when I look at result.explain(), it still says that the wrong index is being used.

However the speed difference is undeniable; so I conclude that 3.5.0 is correctly ignoring the index.

Further testing: If I comment out the line that does the ignore_index(), the query takes 20+ seconds. If I add that line back in to ignore the wrong index that the mysql optimizer chooses then the query takes a few hundred milliseconds.

So it does seem that something's not quite right with whatever 3.6.0 introduced/changed.

Also, is there any way that explain() can correctly show the index that was used? Because I use it to monitor various queries and their times to identify if there needs to be other composite indexes added depending on what people search for.

Thanks for helping!

**Comment by adamchainz** [comment]
> However, when I print result.query, I still don't see the rewritten query - it still has the commented hint in it.

This is expected. The query rewriting is done between django's sql compiler and the database execution, the `.query` property only shows the output from the sql compiler.

> Also when I look at result.explain(), it still says that the wrong index is being used.

What database and version are you using? Possibly upgrading will help here.

> So it does seem that something's not quite right with whatever 3.6.0 introduced/changed.

I am looking into this. Can you check too? The code is in https://github.com/adamchainz/django-mysql/blob/master/src/django_mysql/apps.py#L19 and should be fairly comprehensible.

**Comment by kinghelix** [comment]
> The query rewriting is done between django's sql compiler and the database execution

I suspected this might have been the case, no problem!


>> Also when I look at result.explain(), it still says that the wrong index is being used.
> What database and version are you using? Possibly upgrading will help here.
I'm using MySQL 5.6 on macOS (installed via brew). I'm going to test on a different system with 5.7 just in case.

> I am looking into this. Can you check too?

I'll have another look, but I did already spend a few hours when I was first encountering this issue, digging through that code and the code in query.py etc. I could see the `rewrite_hook` being added to `connection.execute_wrappers` but I found the code in Django which -I think- executed the wrappers and the `connection.execute_wrappers` list was empty at that point, so it never ran anything.

**Comment by adamchainz** [comment]
Fixed in #678, released in version 3.7.1

The problem was for new connection objects created after `AppConfig.ready()` runs. I'm not sure what causes this but managed to emulate it in a test.

**Comment by kinghelix** [comment]
Amazing, thank you so much. I've tested and this is working now.

Incidentally, I upgraded mysql 5.6.44 to 5.7.22 and now the explain is correctly showing the used index (now that the one which the optimizer selected is being ignored).

I really appreciate the quick fix on this!
